### PR TITLE
Enhanced types for Assoc

### PR DESCRIPTION
**Pull request by Harris-Miller**
Updating types for Assoc for better return types when the type for value is the same for the type of the key on the object

Also added tests

### CODE REVIEW

File: test/types/assoc.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import { expectType } from "tsd";
+import { __, assoc } from "../../es/index";
+import { AssocPartialOne} from '../../types/util/tools';
```
Comment (kedashoe) [inline test/types/assoc.ts:3]: I don't believe this is being used?

File: test/types/assoc.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+type Point = { x: number; y: number };
+const p: Point = { x: 2, y: 3 };
+
+const t = assoc('x')(5, p);
```
Comment (kedashoe) [inline test/types/assoc.ts:8]: Does this line do something?

File: types/assoc.d.ts [hunk 3]
```diff
@@ -1,7 +1,10 @@
 import { Placeholder, AssocPartialOne } from './util/tools';
 
-export function assoc<T, U>(__: Placeholder, val: T, obj: U): <K extends string>(prop: K) => Record<K, T> & Omit<U, K>;
-export function assoc<U, K extends string>(prop: K, __: Placeholder, obj: U): <T>(val: T) => Record<K, T> & Omit<U, K>;
-export function assoc<T, U, K extends string>(prop: K, val: T, obj: U): Record<K, T> & Omit<U, K>;
+export function assoc<T, U>(__: Placeholder, val: T, obj: U): <K extends string>(prop: K) => K extends keyof U ? T extends U[K] ? U : Record<K, T> & Omit<U, K> : U & Record<K, T>;
+export function assoc<U, K extends keyof U>(prop: K, __: Placeholder, obj: U): <T>(val: T) => T extends U[K] ? U : Record<K, T> & Omit<U, K>;
+export function assoc<U, K extends string>(prop: K, __: Placeholder, obj: U): <T>(val: T) => U & Record<K, T>;
+export function assoc<K extends keyof U, U>(prop: K, val: U[K], obj: U): U;
```
Comment (kedashoe) [inline types/assoc.d.ts:6]: Is this covered by the next definition? ie `export function assoc<T, K extends keyof U, U>(prop: K, val: T, obj: U): Record<K, T> & Omit<U, K>;`

Is there a reason to have both?

**Review by kedashoe — COMMENTED** [review]
Thanks again for all the PRs! Threw a couple questions inline.

Also, what do you think about copying and pasting, as a comment, the signature you expect each block of tests to hit? So something like
```
// assoc<K extends keyof U, U>(prop: K, val: U[K], obj: U): U;
expectType<Point>(assoc('x', 5, p));
etc..
```

File: test/types/assoc.ts (hunk 2 above)
Comment (Harris-Miller) [reply test/types/assoc.ts:8]: No, I was using it to test, forgot to remove

File: types/assoc.d.ts (hunk 3 above)
Comment (Harris-Miller) [reply types/assoc.d.ts:10]: I looked back at how I wrote this, and with some things I learned from doing my other MRs, I completely rewrote it. It has a lot more comments now so it'll be easier to review. I also was able to remove the `AssocPartialOne` util type, it's no longer needed

File: test/types/assoc.ts (hunk 1 above)
Comment (Harris-Miller) [reply test/types/assoc.ts]: Removed

**Comment by Harris-Miller** [comment]
Closing. Will re-apply changes in new types repo as described here: https://github.com/ramda/ramda/discussions/3329

**Pull request closed without merging**